### DISCUSSION TITLE
toPromise should accept a signal

### DISCUSSION THREAD

**Discussion by boneskull** [statelyai/xstate#5044] [Ideas]
Much like `waitFor()` (see #5042), `toPromise(actor, {signal})` should be a thing.

Upon the `AbortSignal` receiving the `abort` event, `toPromise(actor, {signal})` should:

1. Call `actor.stop()`
2. Remove the `abort` listener
3. Reject with the `reason` prop of the `AbortSignal`

**Comment by Andarist** [comment]
I don't think it should call `.stop`. You can call `toPromise` multiple times - should the cancellation of one of them lead to both being effectively canceled?

**Reply by boneskull** [reply]
Hmm, I didn't think of that.  Maybe it should only call `stop()` if there aren't other subscribers.

**Reply by Andarist** [reply]
Or it shouldn't call `stop` at all 😉 Having this kind of a conditional behavior is a refactoring hazard for consumers

**Reply by boneskull** [reply]
OK, so then would _this_ make sense?

* * *

Upon the AbortSignal receiving the abort event, toPromise(actor, {signal}) should:

1. Remove the `abort` listener
2. Reject with the reason prop of the `AbortSignal`
